Add authorized_keys_path param to simpleserver and client_{private,pu…

simonepelosi opened the pull request:
## Scope and purpose

Fixes #12210 and it's related to #12001 

This PR adds authorized_keys_path param to simpleserver and client_{private,public}_key_path params to simpleclient.

Authorized_keys_path param and client_{private,public}_key_path params help to have a more dynamic configuration for the sshsimpleserver.py and sshsimpleclient.py.
In that way an user can pass their own configuration using the command line without changing the content of the files.

adiroiban committed:
Initial fido2 draft implementation.

adiroiban committed:
Merge branch 'trunk' into 12001-ssh-server-fido2-pubkey-auth

simonepelosi committed:
Add authorized_keys_path param to simpleserver and client_{private,public}_key_path params to simpleclient.

authorized_keys_path param and client_{private,public}_key_path params help to
have a more dynamic configuration for the sshsimpleserver.py and sshsimpleclient.py.
In that way an user can pass their own configuration using the command line without changing
the content of the files.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

jugmac00 commented on docs/conch/examples/sshsimpleclient.py:
```diff
@@ -143,6 +156,20 @@ def closed(self):
+    parser = argparse.ArgumentParser(description="Simple SSH client.")
+
+    parser.add_argument(
+        "--client_public_key_path",
```
I think usually a dash is used for arguments. You can then use the `dest` keyword to convert it to an underscore, though I am not sure whether `argparse` would do that automatically.

jugmac00 commented on docs/conch/examples/sshsimpleserver.py:
```diff
@@ -49,14 +49,6 @@
-sk-ssh-ed25519@openssh.com AAAAGnNrLXNzaC1lZDI1NTE5QG9wZW5zc2guY29tAAAAIKULA+QMG6KFFXeooTZ9dHXfV9wsUzvjBcCHvWJgEYKcAAAAE3NzaDplZDI1NTE5LXl1YmktNWM= user@test
-""".strip()
-
-
```
Would we need to add a file to make the example work again? Or maybe tell the user how to generate the file.

jugmac00 commented on docs/conch/examples/sshsimpleserver.py:
```diff
@@ -300,5 +294,10 @@ def getPrimes(self):
+    parser = argparse.ArgumentParser(description="Simple SSH server.")
+
+    parser.add_argument(
+        "authorized_keys_path", type=str, help="Path to the authorized keys file."
```
Again, I think dashes are used to combine words.

jugmac00 reviewed (commented):
Looks good to me! Though I am not very familiar with the domain.

What about to add some explanation at the top of the module, e.g. how to use the example and also how to generate a new key/file?

adiroiban commented:
Can we have this targeting/merged to `trunk` ?

Right now, we don't have merge process that involved merging a feature branch into another feature branch.

simonepelosi committed:
Handle comments

Using argparse best practices.
Added comment that mention authorized keys file creation.

simonepelosi committed:
Merge branch '12210-add-params-for-authorized-keys-and-client-keys' of github.com:simonepelosi/twisted into 12210-add-params-for-authorized-keys-and-client-keys

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks

for more information, see https://pre-commit.ci

simonepelosi commented:
Target changed and comments addressed!

twm commented on docs/conch/examples/sshsimpleclient.py:
```diff
@@ -66,20 +67,32 @@ def getGenericAnswers(self, name, instruction, questions):
         A deferred can also be returned.
         """
-        return defer.succeed(keys.Key.fromFile(CLIENT_RSA_PRIVATE))
+        args = parser.parse_args()
```
This is deeply weird... you're referencing a `parser` that's conditionally defined in a global scope. And doing that twice?

I don't think that these are appropriate changes to make, as this example deliberately hard-codes things to keep the boilerplate to a minimum.

simonepelosi replied:
Hey! Could you be more specific about `parser` providing me some example? 

Regarding the hard-coded values: maybe having more flexible examples could be better also for twisted devs, we discussed that with @adiroiban https://github.com/twisted/twisted/pull/12002#issuecomment-1971109313

twm replied:
Thank you for providing that context! I'd only read #12210 before reviewing.

Are you aware that you can run a server with `python -m twisted conch`? And that the `conch` script that ships with Twisted is an SSH client?

<details>
<summary>Answers to your questions, but probably irrelevant</summary>

You should parse the parameters in a single location (I suggest defining a `main` function) and thread the values in through `SimpleTransport` down to `SimpleUserAuth`. To do so you'd have to add a `__init__` methods to both `SimpleTransport` and `SimpleUserAuth` that accept those parameters and store them as instance variables. This will add some boilerplate which is presumably why the examples weren't written that way to begin with. Don't rely on global variables.

Unfortunately, these examples are a bit long in the tooth &mdash; `simplesshclient.py` is using `ClientCreator` when it should be using `HostnameEndpoint`. It calls `reactor.run()` when it should use `twisted.internet.task.react()`, etc. You don't have to address that in this PR, but I'd welcome those changes here or in a separate PR. (Really, we need to go through all of the examples and update them.)

Twisted also contains an argument parsing system: `twisted.web.usage.Options`. (I haven't done the archaeology, but `Options` surely predates `optparse` and `argparse` in the stdlib). You could use that, or stick with `argparse` &mdash; I'm indifferent &mdash; but you may be able to find examples if you search for `Options`.

</details>

simonepelosi replied:
@twm thank you for the clarification! I will update this PR with the information you gave me, I ignored the fact that we can run a server using the command you mentioned.

Moreover, for the comments above, I think I should delete this PR and start it again from the trunk branch. 

At this point it makes sense to wait for #12300

adiroiban replied:
I was not aware of the `twisted couch` app.

I am always using the example files so that I can get an idea of what parts are needed and to allow to customize the ssh server.

I guess there are 2 use-cases here. 

* Case A - You just want to get an SSH server up and running for shell access and don't care how it works
* Case B - You want to setup a custom SSH server

For Case A, the `twisted conch` can help... but if you just need a simple/standard SSH server, I am not sure why you would bother with `conch` in the first place :)

For Case B, I think that the example file is much better, especially since there is not much narative documentation for conch ssh.

glyph replied:
> For Case A, the `twisted conch` can help... but if you just need a simple/standard SSH server, I am not sure why you would bother with `conch` in the first place :)

Presumably if someone is working with Conch they will want to try its "out of the box" behavior with the included command line, and it should work properly.  Ideally it should be more configurable so it addresses more of the Case B audience as well, in the same way that e.g. `twist web` can be used to deploy production services

twm reviewed (changes requested):
Hi @simonepelosi. It looks like this is missing test coverage and has some debug code sprinkled about, so I'm going to remove it from the review queue. Please put it back when you're ready for review!

simonepelosi commented:
> Hi @simonepelosi. It looks like this is missing test coverage and has some debug code sprinkled about, so I'm going to remove it from the review queue. Please put it back when you're ready for review!

@adiroiban the code coverage it's related to #12001, should I fix that here ?

adiroiban commented on src/twisted/conch/ssh/keys.py:
```diff
@@ -256,24 +256,38 @@ def _fromString_BLOB(cls, blob):
             )
-        elif keyType == b"ssh-ed25519":
+
+        if keyType == b"sk-ecdsa-sha2-nistp256@openssh.com":
```
I think that this change should be reverted, as this is out of scope for this PR.

the title of this PR is about updating the examples ... but here the main code is modified

simonepelosi replied:
@adiroiban yep, wrongly I created this PR starting from your branch. What about the comment from @twm ? Does it make sense to have this kind of way to pass keys to the examples ?

simonepelosi commented:
I will close this PR and I will open it again starting from the new `trunk` as specified here https://github.com/twisted/twisted/pull/12300#issuecomment-2334173774
